src/Microsoft.Azure.Relay/WebSockets/NetCore21/ClientWebSocket.cs

```diff
@@ -0,0 +1,268 @@
+            }
+
+            // Check that we have not started already
+            var priorState = (InternalState)Interlocked.CompareExchange(ref _state, (int)InternalState.Connecting, (int)InternalState.Created);
+            if (priorState == InternalState.Disposed)
+            {
+                throw new ObjectDisposedException(GetType().FullName);
+            }
+            else if (priorState != InternalState.Created)
+            {
+                throw new InvalidOperationException(SR.net_WebSockets_AlreadyStarted);
+            }
+            _options.SetToReadOnly();
+
+            return ConnectAsyncCore(uri, cancellationToken);
+        }
+
+        private async Task ConnectAsyncCore(Uri uri, CancellationToken cancellationToken)
+        {
+            _innerWebSocket = WebSocketHandle.Create();
+
+            try
+            {
+                // Change internal state to 'connected' to enable the other methods
+                if ((InternalState)Interlocked.CompareExchange(ref _state, (int)InternalState.Connected, (int)InternalState.Connecting) != InternalState.Connecting)
+                {
+                    // Aborted/Disposed during connect.
+                    throw new ObjectDisposedException(GetType().FullName);
+                }
+
+                await _innerWebSocket.ConnectAsyncCore(uri, cancellationToken, _options).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                if (NetEventSource.IsEnabled) { NetEventSource.Error(this, ex); }
+                throw;
+            }
+        }
+
+        public override Task SendAsync(ArraySegment<byte> buffer, WebSocketMessageType messageType, bool endOfMessage,
+            CancellationToken cancellationToken)
+        {
+            ThrowIfNotConnected();
+            return _innerWebSocket.SendAsync(buffer, messageType, endOfMessage, cancellationToken);
+        }
+
+        // Requires .Net Standard 2.1
+        ////public override ValueTask SendAsync(ReadOnlyMemory<byte> buffer, WebSocketMessageType messageType, bool endOfMessage, CancellationToken cancellationToken)
+        ////{
+        ////    ThrowIfNotConnected();
+        ////    return _innerWebSocket.SendAsync(buffer, messageType, endOfMessage, cancellationToken);
+        ////}
+
+        public override Task<WebSocketReceiveResult> ReceiveAsync(ArraySegment<byte> buffer,
+            CancellationToken cancellationToken)
+        {
+            ThrowIfNotConnected();
+            return _innerWebSocket.ReceiveAsync(buffer, cancellationToken);
+        }
+
+        // Requires .Net Standard 2.1
+        ////public override ValueTask<ValueWebSocketReceiveResult> ReceiveAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+        ////{
+        ////    ThrowIfNotConnected();
+        ////    return _innerWebSocket.ReceiveAsync(buffer, cancellationToken);
+        ////}
```

Comment:
Can maybe uncomment these here?

Reply:
I don't feel that I can. These in turn call down to WebSocket methods which are only available in .NET Standard 2.1 Preview.

---
In reply to: [290883715](https://github.com/Azure/azure-relay-dotnet/pull/109#discussion_r290883715) [](ancestors = 290883715)